File: LC198_House Robber_DP.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums: return 0
        dp = [0]*(len(nums)+1)
        dp[1] = nums[0]

        for i in range(2,len(nums)+1):
            dp[i] = max(dp[i-1],dp[i-2]+nums[i-1]) # 数组索引分清，nums取i-1纯粹是索引要求

        return dp[len(nums)]

# reference https://leetcode-cn.com/problems/house-robber/solution/dong-tai-gui-hua-jie-ti-si-bu-zou-xiang-jie-cjavap/
# dp用的是二维还是一维应分清

File: LC198_House Robber_DP.java
class Solution {
    public int rob(int[] nums) {
    if (nums.length == 0) {
        return 0;
    }

    int N = nums.length;
    int[] dp = new int[N+1];
    dp[0] = 0;
    dp[1] = nums[0];
    for (int k = 2; k <= N; k++) {
        dp[k] = Math.max(dp[k-1], nums[k-1] + dp[k-2]);
    }
    return dp[N];
}

}